### PR TITLE
Mutation viewer

### PR DESCRIPTION
**Pull request by kellrott**
- Adds NeedlePlot mutation viewer
 - Makes view tab list generate dynamically
 - Merges auto completion
 - Adds caching for some calls
 - Merges in docker-compose from autocomplete example app

### CODE REVIEW

File: bmeg_app/views/__init__.py [hunk 1]
```diff
@@ -0,0 +1,11 @@
+
+from collections import OrderedDict
+from . import home_view, tumor_match_normal_view, literature_support_view, compare_dresp_view, mutation_view
+
```
Comment (bwalsh) [inline bmeg_app/views/__init__.py:10]: Are these dicts? methods? modules?

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+    && apt-get install openssl \
+    && apt-get install ca-certificates
+
+# this app has an example backend that calls ensemble
+#
+# snafu!: ensemble's certificate has a problem
+# see https://github.com/Ensembl/ensembl-rest/issues/427
+#
+# fix it by downgrading SSL default
+RUN sed -i 's/DEFAULT@SECLEVEL=2/DEFAULT@SECLEVEL=1/' /usr/lib/ssl/openssl.cnf
```
Comment (bwalsh) [inline Dockerfile]: remove this since we are not using ensemble REST

**Review by bwalsh — CHANGES_REQUESTED** [review]

File: bmeg_app/db/__init__.py [hunk 3]
```diff
@@ -1,7 +1,59 @@
...
+        'symbol._3gram',
+        'ensemble_id',
+        'ensemble_id._2gram',
+        'ensemble_id._3gram',
+        ])
+
+    # map
+    # [<Hit(genes/ENSG00000073331): {'ensemble_id': 'ENSG00000073331', 'symbol': 'ALPK1'}>]
+    # to
+    #  {"label": "New York City", "value": "NYC"},
```
Comment (jordan2lee) [inline bmeg_app/db/__init__.py]: appears to be remnants of tutorial. remove these

File: bmeg_app/views/compare_dresp_view.py [hunk 4]
```diff
@@ -20,9 +20,10 @@
 styles=ly.styles
 
 #######
-# Page  
-####### 
-print('loading app layout')   
+# Page
+#######
+print('loading app layout')
+NAME = "Compare Drug Responses"
```
Comment (jordan2lee) [inline bmeg_app/views/compare_dresp_view.py]: branch `layout` has renamed some scripts and text for this view. This `NAME` should reflect the same name as the new tab header

File: bmeg_app/views/home_view.py [hunk 5]
```diff
@@ -20,15 +20,16 @@
 styles=ly.styles
 
 #######
-# Page  
-####### 
+# Page
+#######
+NAME="Main"
```
Comment (jordan2lee) [inline bmeg_app/views/home_view.py]: preference of `"Main"` over `"Home"`?

File: bmeg_app/views/mutation_view.py [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+from plotly.subplots import make_subplots
+import logging
+logger = logging.getLogger(__name__)
+
+#######
+# Prep
+#######
+
+main_colors= ly.main_colors
+styles=ly.styles
```
Comment (jordan2lee) [inline bmeg_app/views/mutation_view.py:24]: Font family and font size should be uniform throughout app, should update this view to match. `styles` dictionary have these values set

File: bmeg_app/views/mutation_view.py [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+  }
+)
+
+
+#######
+# Page
+#######
+NAME="OncoPrint"
+tab_layout = html.Div(children=[
+    html.Label("Gene:"), dcc.Dropdown(id='single-dropdown', value="TP53/ENSG00000141510", search_value="TP53/ENSG00000141510"),
```
Comment (jordan2lee) [inline bmeg_app/views/mutation_view.py]: Match format of other views

`html.Label("Gene"),html.Hr(), dcc.Dropdown(id='single-dropdown', value="TP53/ENSG00000141510", search_value="TP53/ENSG00000141510"),`

File: docker-compose.yml [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+      - elastic
+
+  app:
+    build: ./
+    container_name: bmeg-app
+    # for development: map host directory to container's /app directory
+    volumes:
+    - ./bmeg_app:/app/bmeg_app
+    ports:
+      - 8050:8050
```
Comment (jordan2lee) [inline docker-compose.yml:50]: seems redundant to some contents in `bmeg_app/config.yaml` -- should this be pointed to `'app'['port']` in `bmeg_app/config.yaml`?

Comment (kellrott) [reply docker-compose.yml:50]: Docker needs to know which ports to expose to the outside world

File: bmeg_app/views/__init__.py (hunk 1 above)
Comment (kellrott) [reply bmeg_app/views/__init__.py:10]: Documentation added

File: Dockerfile (hunk 2 above)
Comment (kellrott) [reply Dockerfile]: Done

File: bmeg_app/views/compare_dresp_view.py (hunk 4 above)
Comment (kellrott) [reply bmeg_app/views/compare_dresp_view.py]: This should really be refactored into a config file.

File: bmeg_app/views/home_view.py (hunk 5 above)
Comment (kellrott) [reply bmeg_app/views/home_view.py]: Either way, this part should be refactored into a config file

**Comment by jordan2lee** [comment]
url pathing is disconnected when on `Home` tab and select any of the widgets from the cards (hits else statement for displaying "404: Not found")

File: bmeg_app/db/__init__.py (hunk 3 above)
Comment (kellrott) [reply bmeg_app/db/__init__.py]: done

**Comment by kellrott** [comment]
> url pathing is disconnected when on `Home` tab and select any of the widgets from the cards (hits else statement for displaying "404: Not found")

Fixed, but those sections should probably be dynamically generated rather then hard coded

File: bmeg_app/views/mutation_view.py (hunk 7 above)
Comment (kellrott) [reply bmeg_app/views/mutation_view.py]: Updated

File: bmeg_app/views/compare_dresp_view.py [hunk 9]
```diff
@@ -21,12 +21,36 @@
...
+                            dbc.ModalBody( '• Dive deeper to see the characteristics of the samples that generated the violin plots of particular drugs.'),
+                            dbc.ModalFooter(dbc.Button('Close',id='close1',className='ml-auto')),
+                        ],
+                        id='main_help1',
+                        size='med',
+                        centered=True,
+                    ),
+                ]),
+                width=1,
+            ),
```
Comment (jordan2lee) [inline bmeg_app/views/compare_dresp_view.py]: remove these lines. old code that has been moved elsewhere now. causes duplicate `id` that breaks callback

**Review by jordan2lee — CHANGES_REQUESTED** [review]

File: bmeg_app/views/compare_dresp_view.py (hunk 4 above)
Comment (jordan2lee) [reply bmeg_app/views/compare_dresp_view.py]: for now hardcode `NAME = "Cancer Drug Screening"` to match with name on Home page

File: bmeg_app/views/compare_dresp_view.py (hunk 9 above)
Comment (kellrott) [reply bmeg_app/views/compare_dresp_view.py]: Removed.

File: bmeg_app/views/mutation_view.py (hunk 6 above)
Comment (kellrott) [reply bmeg_app/views/mutation_view.py:24]: Done.

**Review by jordan2lee — APPROVED** [review]

**Comment by jordan2lee** [comment]
looks good

**Pull request merged**